### PR TITLE
GP regression with derivative information

### PR DESCRIPTION
**Pull request by dme65**
Adding a new kernel `RBFKernelGrad` and a mean `ConstantMeanGrad` that can be used together with MultiTask to train GPs from value and derivative information.

### CODE REVIEW

File: gpytorch/kernels/scale_kernel.py [hunk 1]
```diff
@@ -94,11 +94,12 @@ def forward(self, x1, x2, batch_dims=None, diag=False, **params):
...
-
         if torch.is_tensor(orig_output):
             outputscales = outputscales.view(-1, *([1] * (orig_output.dim() - 1)))
 
         if diag:
             return delazify(orig_output) * outputscales
-
         return orig_output.mul(outputscales)
+
+    def size(self, x1, x2):
```
Comment (gpleiss) [inline gpytorch/kernels/scale_kernel.py]: We might want to add this to our other decorator kernels as well (GridInterpolationKernel, ProductKernel, etc.)

File: gpytorch/means/constant_mean_grad.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env python3
+
+import torch
+from .mean import Mean
+
+
+class ConstantMeanGrad(Mean):
```
Comment (gpleiss) [inline gpytorch/means/constant_mean_grad.py]: Is there a way that we could make this more general - like how `MultitaskMean` wraps arbitrary means?

File: gpytorch/kernels/rbf_kernel_grad.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        b = 1
+        if len(x1.size()) == 2:
+            n1, d = x1.size()
+            n2, d = x2.size()
+        else:
+            b, n1, d = x1.size()
+            _, n2, _ = x2.size()
+
+        K = torch.zeros(b, n1 * (d + 1), n2 * (d + 1), device=x1.device, dtype=x1.dtype)  # batch x n1(d+1) x n2(d+1)
+        ell = self.lengthscale.squeeze()
```
Comment (gpleiss) [inline gpytorch/kernels/rbf_kernel_grad.py]: This squeeze will cause broadcasting issues if `batch_size != 1`. Is it necessary?

Comment (dme65) [reply gpytorch/kernels/rbf_kernel_grad.py]: I get a shape error when I call `likelihood(model(test_x)).confidence_region()` if I remove this squeeze. Is there an obvious reason why?

File: gpytorch/means/constant_mean_grad.py (hunk 2 above)
Comment (dme65) [reply gpytorch/means/constant_mean_grad.py]: I'm not sure if this is possible. The mean vector evaluated at one point is a vector of length d+1 with the original scalar mean function and the corresponding d partial derivatives along x_1, ..., x_d. I think this class has to be implemented for each mean, just like ConstantMean etc.

Comment (gpleiss) [reply gpytorch/means/constant_mean_grad.py]: Okay that should be fine then.

**Pull request merged**